#Connect to Etherium Node

#Activity #1: Setup the environment

In [1]:
!pip install protobuf==3.20.3
!pip install web3==5.31.4

  Using cached protobuf-3.20.3-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.1 MB)
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.19.5
    Uninstalling protobuf-3.19.5:
      Successfully uninstalled protobuf-3.19.5
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
web3 5.31.4 requires protobuf==3.19.5, but you have protobuf 3.20.3 which is incompatible.


  Using cached protobuf-3.19.5-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.1 MB)
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.20.3
    Uninstalling protobuf-3.20.3:
      Successfully uninstalled protobuf-3.20.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorboard 2.14.1 requires protobuf>=3.19.6, but you have protobuf 3.19.5 which is incompatible.
tensorflow 2.14.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.19.5 which is incompatible.
tensorflow-datasets 4.9.3 requires protobuf>=3.20, but you have protobuf 3.19.5 which is incompatible.
tensorflow-hub 0.15.0 requires protobuf>=3.19.6, but you have protobuf 3.19.5 which is incompatible.
tensorflow-metadata 1.14.0 requires protobuf<4.21,>=3.20.3, but you have protobuf 3.19.5 which is in

In [3]:
from web3 import Web3

#Activity #2-3: Connect to an Ethereum node (aka “provider”)

In [4]:
providerURL = 'https://mainnet.infura.io/v3/7b5b98d59854437994457dd5791ba80f'
w3 = Web3(Web3.HTTPProvider(providerURL))

In [6]:
w3.isConnected()

True

#Activity #4:Interact with the Ethereum blockchain

In [8]:
providerURL = 'https://mainnet.infura.io/v3/7b5b98d59854437994457dd5791ba80f'
w3 = Web3(Web3.HTTPProvider(providerURL))
latestBlock = w3.eth.get_block('latest')
print(latestBlock)

AttributeDict({'baseFeePerGas': 14762972052, 'difficulty': 0, 'extraData': HexBytes('0x496c6c756d696e61746520446d6f63726174697a6520447374726962757465'), 'gasLimit': 30000000, 'gasUsed': 15401163, 'hash': HexBytes('0xcf9976bed4b045d86ff100a7716a26def1e2e3acad417a8214fc98503f18c3e0'), 'logsBloom': HexBytes('0xf03b05eecfcfe008937b6217982930e950839835298e28f03ed93003be0223a00c2591808d27963d0167d620061d0bb103a8ba1caa4eb9acec13752ca5e221ade258ce18490eed6ee9f3d71e54e222ac8693140b94f6498f0d84fce88c689651df534e0e8ed7e901b58c531074023e7d26f800e056bc0e548fe0531458ca5e967643a2d286118f282fe7a3f74b1a10cfd8c31a239989114c686bbc6c229174b89b348168c78a79572f544de51b1824c21a8c2125cc2f885b248d40f2d4320521f4d0056627164d4c870890234f9a12e58fb1b62b752f089e836ee1cb053be80d4cf9f40c20919ec9c54511c878048862c1a19d518bfe494405034f280ae5d931'), 'miner': '0xDAFEA492D9c6733ae3d56b7Ed1ADB60692c98Bc5', 'mixHash': HexBytes('0x221c87ecf6257e26a2acbb9baa95ac5ee8caa87c6499bd2ab10b10ff6cb09b92'), 'nonce': HexBytes('0x00000000

The format of my result is: AttributeDict({'baseFeePerGas': 14762972052, 'difficulty': 0, 'extraData': HexBytes( .... )

 We observe information about the latest block on the Ethereum mainnet, the key attributes in this object are:

    baseFeePerGas: This attribute represents the base fee per unit of gas used in Ethereum transactions. It's a measure of the current cost of executing transactions on the Ethereum network.

    difficulty: The current level of difficulty for mining new blocks on the Ethereum blockchain. It's a measure of how hard it is to find a valid block.

    extraData: Extra data included in the Ethereum block, which can vary depending on the network or miner.

    gasLimit: The maximum amount of gas that can be used in a block.

    gasUsed: The total amount of gas used by all transactions included in the block.

    hash: The unique identifier of the block, often referred to as the block hash.

    logsBloom: A data structure representing the Bloom filter of the logs in the block.

    miner: The address of the miner who successfully mined this block.

    mixHash: A hash value related to the proof-of-work algorithm used in mining.

    nonce: A random number used in the mining process.

    number: The block number, which is a unique identifier for each block in the Ethereum blockchain.

    parentHash: The hash of the parent block.

    receiptsRoot: The root hash of the transaction receipts included in the block.

    sha3Uncles: The SHA3 hash of the stale blocks of the current block.

    size: The size of the block in bytes.

    stateRoot: The root hash of the Ethereum state trie after applying all transactions.

    timestamp: The timestamp when the block was mined, in Unix timestamp format.

    totalDifficulty: The cumulative difficulty of the chain up to this block.

    transactions: A list of transactions included in the block.

These attributes provide information about a latest block in the Ethereum blockchain and are used for tracking and analyzing blockchain data. These attributes can be used for various purposes, such as monitoring network health, analyzing transaction costs, and understanding the blockchain's history.

In [9]:
wallet = w3.toChecksumAddress('0xDAFEA492D9c6733ae3d56b7Ed1ADB60692c98Bc5')
print(w3.eth.get_balance(wallet))

9060276572462783414


value in wei: 9060276572462783414

In [10]:
# Value in Wei
wei_value = 9060276572462783414

# Convert Wei to Ether
ether_value = w3.fromWei(wei_value, 'ether')

print(f"Value in Ether: {ether_value}")

Value in Ether: 9.060276572462783414


Value in Ether: 9.060276572462783414

#Activity #5: Interact with a smart contract on Ethereum blockchain (Mainnet)

In [13]:
abi = '[{"inputs":[{"internalType":"string","name":"name","type":"string"},{"internalType":"string","name":"symbol","type":"string"},{"internalType":"address","name":"baycAddress","type":"address"},{"internalType":"address","name":"baccAddress","type":"address"}],"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"owner","type":"address"},{"indexed":true,"internalType":"address","name":"approved","type":"address"},{"indexed":true,"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"Approval","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"owner","type":"address"},{"indexed":true,"internalType":"address","name":"operator","type":"address"},{"indexed":false,"internalType":"bool","name":"approved","type":"bool"}],"name":"ApprovalForAll","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"uint256","name":"_currentPrice","type":"uint256"},{"indexed":true,"internalType":"uint256","name":"_timeElapsed","type":"uint256"}],"name":"MutantPublicSalePaused","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"uint256","name":"_saleDuration","type":"uint256"},{"indexed":true,"internalType":"uint256","name":"_saleStartTime","type":"uint256"}],"name":"MutantPublicSaleStart","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"previousOwner","type":"address"},{"indexed":true,"internalType":"address","name":"newOwner","type":"address"}],"name":"OwnershipTransferred","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"uint256","name":"_mintedMutantsStartingIndex","type":"uint256"},{"indexed":true,"internalType":"uint256","name":"_megaMutantsStartingIndex","type":"uint256"}],"name":"StartingIndicesSet","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"from","type":"address"},{"indexed":true,"internalType":"address","name":"to","type":"address"},{"indexed":true,"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"Transfer","type":"event"},{"inputs":[],"name":"MAYC_PROVENANCE","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"NUM_MEGA_MUTANTS","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"PS_MAX_MUTANTS","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"PS_MAX_MUTANT_PURCHASE","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"PS_MUTANT_ENDING_PRICE","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"SERUM_MUTATION_OFFSET","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"approve","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"owner","type":"address"}],"name":"balanceOf","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"collectionStartingIndexBlock","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"getApproved","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getMintPrice","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"apeId","type":"uint256"},{"internalType":"uint8","name":"serumTypeId","type":"uint8"}],"name":"getMutantIdForApeAndSerumCombination","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getRemainingSaleTime","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint8","name":"serumType","type":"uint8"},{"internalType":"uint256","name":"apeId","type":"uint256"}],"name":"hasApeBeenMutatedWithType","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"owner","type":"address"},{"internalType":"address","name":"operator","type":"address"}],"name":"isApprovedForAll","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"isMinted","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"megaMutantsStartingIndex","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"numMutants","type":"uint256"}],"name":"mintMutants","outputs":[],"stateMutability":"payable","type":"function"},{"inputs":[],"name":"mintedMutantsStartingIndex","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"serumTypeId","type":"uint256"},{"internalType":"uint256","name":"apeId","type":"uint256"}],"name":"mutateApeWithSerum","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"name","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"numMutantsMinted","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"owner","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"ownerOf","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"pausePublicSale","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"publicSaleActive","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"publicSaleDuration","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"publicSaleMutantStartingPrice","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"publicSaleStartTime","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"renounceOwnership","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"safeTransferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"},{"internalType":"bytes","name":"_data","type":"bytes"}],"name":"safeTransferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"serumMutationActive","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"operator","type":"address"},{"internalType":"bool","name":"approved","type":"bool"}],"name":"setApprovalForAll","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"string","name":"uri","type":"string"}],"name":"setBaseURI","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"setStartingIndices","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"saleDuration","type":"uint256"},{"internalType":"uint256","name":"saleStartPrice","type":"uint256"}],"name":"startPublicSale","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes4","name":"interfaceId","type":"bytes4"}],"name":"supportsInterface","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"symbol","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"togglePublicSaleActive","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"toggleSerumMutationActive","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"index","type":"uint256"}],"name":"tokenByIndex","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"owner","type":"address"},{"internalType":"uint256","name":"index","type":"uint256"}],"name":"tokenOfOwnerByIndex","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"tokenURI","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"totalApesMutated","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"totalSupply","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"transferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"newOwner","type":"address"}],"name":"transferOwnership","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"withdraw","outputs":[],"stateMutability":"nonpayable","type":"function"}]'
contractAddress = '0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D'
#Connect to remote provider
w3 = Web3(Web3.HTTPProvider(providerURL))
#Check if connected
w3.isConnected()
#Create an instance of the contract
contractInstance = w3.eth.contract(address = contractAddress, abi = abi)
print(contractInstance)

In [14]:
# call the function to get totalSupply
totalSupply = contractInstance.functions.totalSupply().call()
print(totalSupply)

10000


In [17]:
    # Task-01: The owner address of the contract
    owner = contractInstance.functions.owner().call()
    print(f'Task-01: Owner Address of the Contract: {owner}')

    # Task-02: The total supply of tokens under this collection
    totalSupply = contractInstance.functions.totalSupply().call()
    print(f'Task-02: Total Supply of Tokens: {totalSupply}')

    # Task-03: The symbol of the token collection
    symbol = contractInstance.functions.symbol().call()
    print(f'Task-03: Token Collection Symbol: {symbol}')

    # Task-04: The token URI of Bored Ape NFT token with id 2321
    tokenURI = contractInstance.functions.tokenURI(2321).call()
    print(f'Task-04: Token URI for Token ID 2321: {tokenURI}')

    # Task-05: The token owner of Bored Ape NFT token with id 2321
    tokenOwner = contractInstance.functions.ownerOf(2321).call()
    print(f'Task-05: Owner of Token ID 2321: {tokenOwner}')

Task-01: Owner Address of the Contract: 0x0000000000000000000000000000000000000000
Task-02: Total Supply of Tokens: 10000
Task-03: Token Collection Symbol: BAYC
Task-04: Token URI for Token ID 2321: ipfs://QmeSjSinHpPnmXmspMjwiXyN6zS4E9zccariGR3jxcaWtq/2321
Task-05: Owner of Token ID 2321: 0xe3199072644455D19f58B1fd8106Ac80b3d2e780


Results: Task-01: Owner Address of the Contract: 0x0000000000000000000000000000000000000000
Task-02: Total Supply of Tokens: 10000
Task-03: Token Collection Symbol: BAYC
Task-04: Token URI for Token ID 2321: ipfs://QmeSjSinHpPnmXmspMjwiXyN6zS4E9zccariGR3jxcaWtq/2321
Task-05: Owner of Token ID 2321: 0xe3199072644455D19f58B1fd8106Ac80b3d2e780